### PR TITLE
Fully destructure slice and array constants into patterns

### PR DESCRIPTION
**Pull request by oli-obk**
There were a lot of workarounds for making such constants not opaque, now we expand all constants into patterns unless they are not structural match

### CODE REVIEW

**Comment by oli-obk** [comment]
ok, this is a perf improvement across the board. Unless people start putting `include_bytes!(some_humongous_file)` into constants used in patterns we're going to be fine. And if they do, we can always add an arbitrary limit after which we treat a slice/array constant as opaque

ready for review @varkor

File: compiler/rustc_mir_build/src/thir/pattern/_match.rs [hunk 1]
```diff
@@ -2207,19 +2111,7 @@ fn pat_constructor<'tcx>(
             if let Some(int_range) = IntRange::from_const(tcx, param_env, value, pat.span) {
```
Comment (RalfJung) [inline compiler/rustc_mir_build/src/thir/pattern/_match.rs:2111]: Longer-term, would it make sense to make `PatKind::Constant` fully opaque, which I guess means not using it for integer ranges either?

File: compiler/rustc_mir_build/src/thir/pattern/_match.rs [hunk 2]
```diff
@@ -2207,19 +2111,7 @@ fn pat_constructor<'tcx>(
...
-                    }
-                    (ty::ConstKind::Value(ConstValue::Slice { start, end, .. }), ty::Slice(_)) => {
-                        let len = (end - start) as u64;
-                        Some(Slice(Slice { array_len: None, kind: FixedLen(len) }))
-                    }
-                    // FIXME(oli-obk): implement `deref` for `ConstValue`
-                    // (ty::ConstKind::Value(ConstValue::ByRef { .. }), ty::Slice(_)) => { ... }
-                    _ => Some(ConstantValue(value)),
-                }
+                Some(ConstantValue(value))
```
Comment (RalfJung) [inline compiler/rustc_mir_build/src/thir/pattern/_match.rs:2114]: Is this related to exhaustiveness checking? Shouldn't opaque consts be treated as matching nothing for that as they might not be structural-match?

File: compiler/rustc_mir_build/src/thir/pattern/_match.rs (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs:2111]: that won't work, as we have no way to distinguish literal integers in patterns from constants.

Comment (RalfJung) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs:2111]: I thought structural match things like integers end up with a different `PatKind` that `Constant`?

Comment (oli-obk) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs:2111]: They do not (yet). That is an independent refactoring that we should analyze in isolation on its effect on readability, diagnostics and perf.

Comment (RalfJung) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs:2111]: Ah I see.

File: compiler/rustc_mir_build/src/thir/pattern/_match.rs (hunk 2 above)
Comment (oli-obk) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs]: it is, very good catch. We still need a variant for string slices here, as string slices are not opaque, but `ConstantValue` is gone now and the replacement `Str` only handles `&str` constants. There are further improvements we can do here now that everything is explicit, but I'd rather not touch `compare_const_vals` in this PR.

Comment (RalfJung) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs]: I should note that I understand basically nothing of this code, if I caught anything that was an accident.^^

File: compiler/rustc_mir_build/src/thir/pattern/_match.rs [hunk 3]
```diff
@@ -820,8 +820,8 @@ enum Constructor<'tcx> {
     Single,
     /// Enum variants.
     Variant(DefId),
-    /// Literal values.
-    ConstantValue(&'tcx ty::Const<'tcx>),
+    /// String literals
+    Str(&'tcx ty::Const<'tcx>),
```
Comment (RalfJung) [inline compiler/rustc_mir_build/src/thir/pattern/_match.rs:824]: Okay so this type is indeed only used for exhaustiveness checking, that's why we do not need consts in here? Maybe add that to the type doccomment.

File: compiler/rustc_mir_build/src/thir/pattern/_match.rs (hunk 2 above)
Comment (RalfJung) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs]: Also if the last 2 comments are bugfixes, is it possible to add a test?

Comment (oli-obk) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs]: There was no bugfix, no behaviour was changed, all I did was to move the bail-out (that was guaranteed to happen later) to an earlier site so the rest of the code that I was able to remove became dead code. This PR generally is mostly just removing dead code that became obsolete after destructuring all structural-eq constants into (non-const) patterns. This means except for `&str`, all `PatKind::Const` are now opaque

Comment (RalfJung) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs]: I see. Maybe add some doc comments to `PatKind` explaining this. :)

File: compiler/rustc_mir_build/src/thir/pattern/_match.rs (hunk 3 above)
Comment (varkor) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs:824]: Why are strings special here?

**Comment by varkor** [comment]
Once there's a comment explaining why strings are special, this should be good.
cc @Nadrieril, who might have comments.

**Comment by RalfJung** [comment]
> Once there's a comment explaining why strings are special, this should be good.
> cc @Nadrieril, who might have comments.

I hope Oli can also add a comment [clarifying this](https://github.com/rust-lang/rust/pull/77390#discussion_r499502213). :)

File: compiler/rustc_mir_build/src/thir/pattern/_match.rs [hunk 4]
```diff
@@ -820,7 +820,11 @@ enum Constructor<'tcx> {
     Single,
     /// Enum variants.
     Variant(DefId),
-    /// String literals
+    /// String literals. These have their own variant, because they behave essentially like
+    /// `&[u8]`, but there's no way to pattern match on parts of strings. So we could convert them
+    /// to slice constructors, but then we'd have to special case other code sites to not treat them
+    /// *exactly* like slice constructors. It seems simpler overall to just give them their own
+    /// variant.
```
Comment (RalfJung) [inline compiler/rustc_mir_build/src/thir/pattern/_match.rs:827]: Is it true that `Constructor` is used only for exhaustiveness checking, but does not actually effect MIR construction for pattern matching?

If yes, I think it would be good to mention that in the `enum Constructor` doc comment (but doesn't have to be in this PR).

Comment (oli-obk) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs:827]: I don't actually know. Decisions on MIR match construction may be made on information obtained via `Constructor`.

File: compiler/rustc_mir_build/src/thir/pattern/_match.rs [hunk 5]
```diff
@@ -2110,7 +2101,12 @@ fn pat_constructor<'tcx>(
             if let Some(int_range) = IntRange::from_const(tcx, param_env, value, pat.span) {
                 Some(IntRange(int_range))
             } else {
-                Some(ConstantValue(value))
+                match value.ty.kind() {
+                    ty::Ref(_, t, _) if t.is_str() => Some(Str(value)),
+                    ty::Float(_) => Some(FloatRange(value, value, RangeEnd::Included)),
+                    // Non structural-match values are opaque.
+                    _ => None,
```
Comment (Nadrieril) [inline compiler/rustc_mir_build/src/thir/pattern/_match.rs:2112]: `None` here explicitely means a wildcard, so I'm not sure this is correct. Does this behave correctly? I would expect this implies we'd consider a branch that uses an opaque pattern to exhaustively match anything, which is the opposite of how it should behave.
What's an example of such an opaque constant? I'd like to construct a test.

I would instead add an `Opaque` constructor for this case to force us to treat it correctly. (I've also been meaning to add `Wildcard` as a constuctor instead of returning an `Option` here).

**Comment by Nadrieril** [comment]
Omg yes you removed `slice_pat_covered_by_const`! I hated that function. I'm happy about this, const handling was a major source of complexity.
If I understand correctly, what you're doing is removing code that is now dead since https://github.com/rust-lang/rust/pull/70743? I haven't read that PR in enough detail to check if all the code you're removing is inaccessible.

File: compiler/rustc_mir_build/src/thir/pattern/_match.rs [hunk 6]
```diff
@@ -1485,9 +1485,7 @@ fn all_constructors<'a, 'tcx>(
         )
     };
     match *pcx.ty.kind() {
-        ty::Bool => {
-            [true, false].iter().map(|&b| ConstantValue(ty::Const::from_bool(cx.tcx, b))).collect()
-        }
+        ty::Bool => vec![make_range(0, 1)],
         ty::Array(ref sub_ty, len) if len.try_eval_usize(cx.tcx, cx.param_env).is_some() => {
```
Comment (Nadrieril) [inline compiler/rustc_mir_build/src/thir/pattern/_match.rs:1484]: Doesn't this affect diagnostics? I'd expect this would display "pattern `0` not covered" instead of "pattern `false` not covered".

**Comment by oli-obk** [comment]
> If I understand correctly, what you're doing is removing code that is now dead since #70743? I haven't read that PR in enough detail to check if all the code you're removing is inaccessible.

Well, I also made a small change here (`&[u8]` and `&[u8; N]` constants get converted into slice patterns, thus leaving *no* constants but `&str` or opaque constants as `PatKind::Const`). This allowed all of that code to become dead, but parts of it were already dead after #70743

File: compiler/rustc_mir_build/src/thir/pattern/_match.rs (hunk 4 above)
Comment (Nadrieril) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs:827]: Yes, it's only for exhaustiveness

Comment (Nadrieril) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs:827]: I know because I've changed Constructor without any code breaking outside of this `_match.rs`.

File: compiler/rustc_mir_build/src/thir/pattern/_match.rs (hunk 5 above)
Comment (oli-obk) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs:2112]: I remember trying something like this, and could not figure out all the details. If you want to try to add an `Opaque` constructor, it's easy to construct an opaque constant by using a type that doesn't derive `PartialEq` but manually implements it:

```rust
#[derive(PartialEq)]
struct Opaque(i32);

impl Eq for Opaque {}

const FOO: &&Opaque = &&Opaque(42);

match FOO {
    FOO => {},
    // The following should not lint about unreachable
    Opaque(42) => {}
}
```

amusingly the above ICEs on the nightly playground, not sure if that has my PR yet.

File: compiler/rustc_mir_build/src/thir/pattern/_match.rs (hunk 6 above)
Comment (varkor) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs:1484]: I had also wondered this, but assumed this would be caught by tests, so it must have been okay. But you're right that it would be better to understand this (and add a comment).

File: compiler/rustc_mir_build/src/thir/pattern/_match.rs (hunk 5 above)
Comment (RalfJung) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs:2112]: >  a type that doesn't derive PartialEq but manually implements it:

But this derives it, no manual impl? Or did you mean `Eq`?

File: compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs [hunk 7]
```diff
@@ -390,11 +390,14 @@ impl<'a, 'tcx> ConstToPat<'a, 'tcx> {
                 ty::Slice(elem_ty) if elem_ty == tcx.types.u8 => PatKind::Constant { value: cv },
```
Comment (Nadrieril) [inline compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs]: Why are we still converting some slices to `PatKind::Constant`? Aren't the only constants left `&str` and opaque ones?

**Comment by varkor** [comment]
I'm concerned by https://github.com/rust-lang/rust/pull/77390#discussion_r503370508. I think we should delay merging this until @Nadrieril's concerns have been addressed.

@bors r-

File: compiler/rustc_mir_build/src/thir/pattern/_match.rs (hunk 6 above)
Comment (oli-obk) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs:1484]: This is because while *decisions* are made on `Constructor`s, they do not directly flow into diagnostics. Instead diagnostics are reported on `Pat`s in https://github.com/rust-lang/rust/blob/6553d38bcb0c9193808ae93db8bd95bcf4c4824a/compiler/rustc_mir_build/src/thir/pattern/check_match.rs

File: compiler/rustc_mir_build/src/thir/pattern/_match.rs (hunk 5 above)
Comment (oli-obk) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs:2112]: I think either will do, we require both for structural equality. I was just lazy and didn't want to write the impl when the `Eq` impl is trivial to write

File: compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs (hunk 7 above)
Comment (oli-obk) [reply compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs]: You're looking on something that is changed in a later commit

**Comment by Nadrieril** [comment]
> Well, I also made a small change here (`&[u8]` and `&[u8; N]` constants get converted into slice patterns, thus leaving _no_ constants but `&str` or opaque constants as `PatKind::Const`). This allowed all of that code to become dead, but parts of it were already dead after #70743

Wait, integer and float ranges are still `PatKind::Const`, right?

File: compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs (hunk 7 above)
Comment (Nadrieril) [reply compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs]: Oops sorry

**Comment by oli-obk** [comment]
> Wait, integer and float ranges are still `PatKind::Const`, right?

hmm yes, oops, forgot about those. Booleans and chars, too.

File: compiler/rustc_mir_build/src/thir/pattern/mod.rs [hunk 8]
```diff
@@ -158,6 +158,9 @@ crate enum PatKind<'tcx> {
         subpattern: Pat<'tcx>,
     },
 
+    /// Opaque constant. This will be treated as a single, but unknown value.
+    /// Except for `&str`, which will be handled as a string pattern and thus exhaustiveness
+    /// checking will detect if you use the same string twice in different patterns.
     Constant {
```
Comment (Nadrieril) [inline compiler/rustc_mir_build/src/thir/pattern/mod.rs:164]: This comment should also mention in ranges etc then

**Comment by RalfJung** [comment]
> I'm concerned by [#77390 (comment)](https://github.com/rust-lang/rust/pull/77390#discussion_r503370508). I think we should delay merging this until @Nadrieril's concerns have been addressed.

CI is already running, so you also need to
@bors retry

File: compiler/rustc_mir_build/src/thir/pattern/_match.rs (hunk 5 above)
Comment (Nadrieril) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs:2112]: Your example ICEs because of the refs, that's been an open issue for a while.
```rust
#[derive(PartialEq)]
struct Opaque(i32);

impl Eq for Opaque {}

const FOO: Opaque = Opaque(42);
match FOO {
    FOO => {},
    // The following should not lint about unreachable
    Opaque(42) => {}
}
```
This works, but does report the second branch as unreachable on both stable and nightly. I don't understand why, and I haven't seen it reported as a bug yet.

Comment (Nadrieril) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs:2112]: The good thing is that your incorrect usage of `None` above won't be a regression x)

Comment (oli-obk) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs:2112]: I did an experiment by returning ` Some(NonExhaustive)`, which made patterns like `FOO` above lint that they are unreachable, so that's definitely not it. I'm ot sure what `Constructor::Opaque` would do differently from `NonExhaustive`, but I'm going through all the uses of `NonExhaustive` now to analyze that

Comment (Nadrieril) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs:2112]: A simpler way to know what the `Opaque` constructor should do is look at all the places where `ConstantValue` was used but the constant wasn't inspected

Comment (Nadrieril) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs:2112]: Actually, an easy way to avoid a regression would be that instead of removing `ConstantValue` altogether like you did, you would rename it to `Opaque` and see what's left.

Comment (Nadrieril) [reply compiler/rustc_mir_build/src/thir/pattern/_match.rs:2112]: All in all, given that matching with opaque constants is an error anyways, this will not allow any unsound code. So I'm not too worried if we let that slip. We can file the above as a bug and I'll introduce `Opaque` myself whenever I get some time and a powerful enough computer to play with rustc

**Comment by oli-obk** [comment]
@Nadrieril can you help me with the latest commit? I have absolutely no clue how to create a `Constructor::Opaque` that will work. Everything I read and tested led me to believe the last commit should work, but instead it still ICEs the example I showed you and treats raw pointer patterns like wildcard patterns

**Comment by Nadrieril** [comment]
So, your example ICEs in stable ([playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b6ea99f29c06a5c9c9fb67bc63bffdb0)), so that's probably unrelated to the current PR.
Then, if I remove the double ref it doesn't ICE but emits an "unreachable pattern" warning that's clearly incorrect, again on stable ([playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ac9dbcfd6f52979f3fb29be7087cde82)).
So the problems may not come from you getting the `Opaque` constructor wrong. I'll have a look tomorrow or this week-end and see if I understand what's happening

**Comment by Nadrieril** [comment]
I'm pretty sure I found how to fix one of the bugs: see https://github.com/rust-lang/rust/issues/78057 . I'm not familiar enough with `const-to-pat.rs` to fix it myself.
I'm still going through your PR to try to make `Opaque` work

**Comment by Nadrieril** [comment]
I was struggling to realize why I couldn't make `Opaque` work, and I've found why: if I replace every construction of `Opaque` with a `panic!()`, it appears that it's not constructed at all for most of the cases I was testing! In fact const_to_pat expands constants to `PatKind::Wild` in many error cases like https://github.com/rust-lang/rust/blob/6af9846fcc8797bf97e9fb387385208c2219f3d0/compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs#L337-L348
I don't like these wildcards. Wrt `_match.rs`, I'd like to emit `PatKind::Constant` instead. Would that break anything? Who else uses the output of `const_to_pat`?

**Comment by Nadrieril** [comment]
@oli-obk I fixed `Opaque`! I submitted https://github.com/rust-lang/rust/pull/78072 that contains your changes plus mine.
I also filed two issues #78071 and #78057 from the discussion we had here. They are not regressions nor directly related to the current PR so I thought we'd deal with them separately.

**Comment by oli-obk** [comment]
closing in favour of #78072

Thanks so much!
